atc/metric/emitter/newrelic.go

```diff
@@ -173,16 +144,117 @@ func (emitter *NewRelicEmitter) Emit(logger lager.Logger, event metric.Event) {
 	// otherwise recording it. (This won't be easily graphable, that's okay,
 	// this is more for monitoring synthetics)
 	if event.State != metric.EventStateOK {
-		singlePayload := emitter.simplePayload(logger, event, "alert")
+		singlePayload := emitter.transformToNewRelicEvent(logger, event, "alert")
 		// We don't have friendly names for all the metrics, and part of the
 		// point of this alert is to catch events we should be logging but
 		// didn't; therefore, be consistently inconsistent and use the
 		// concourse metric names, not our translation layer.
 		singlePayload["metric"] = event.Name
-		payload = append(payload, singlePayload)
+		emitter.NewRelicBatch = append(emitter.NewRelicBatch, singlePayload)
+	}
+
+	duration := time.Since(emitter.LastEmitTime)
+	if len(emitter.NewRelicBatch) >= emitter.BatchSize || duration >= emitter.BatchDuration {
+		logger.Debug("pre-emit-batch", lager.Data{
+			"batch-size":         emitter.BatchSize,
+			"current-batch-size": len(emitter.NewRelicBatch),
+			"batch-duration":     emitter.BatchDuration,
+			"current-duration":   duration,
+		})
+		emitter.submitBatch(logger)
+	}
+}
+
+// NewRelic has strict requirements around the structure of the events
+// Keys must be alphanumeric and can contain hyphens or underscores
+// Values must be sting, int, or unix timestamps. No maps/arrays.
+func (emitter *NewRelicEmitter) transformToNewRelicEvent(logger lager.Logger, event metric.Event, nameOverride string) NewRelicEvent {
+	name := nameOverride
+	if name == "" {
+		name = strings.Replace(event.Name, " ", "_", -1)
+	}
+
+	eventType := fmt.Sprintf("%s%s", emitter.prefix, name)
```

Comment:
I'd go with a plain `+` here rather than `Sprintf` (which would do all of the formatting work) as it'd not look terrible

```
BenchmarkFormatting-12           6101223               196 ns/op
BenchmarkAddition-12            13326642                87.4 ns/op
```

```go
func BenchmarkFormatting(b *testing.B) {
	var res string
	for n := 0; n < b.N; n++ {
		foo := strconv.Itoa(b.N)
		bar := strconv.Itoa(b.N)

		res = fmt.Sprintf("%s%s", foo, bar)
	}

	fmt.Fprintln(ioutil.Discard, res)
}

func BenchmarkAddition(b *testing.B) {
	var res string
	for n := 0; n < b.N; n++ {
		foo := strconv.Itoa(b.N)
		bar := strconv.Itoa(b.N)

		res = foo + bar
	}

	fmt.Fprintln(ioutil.Discard, res)
}
```

Reply:
That's fine. This is the old code, it just looks new because we moved the functions around.